src/org/apache/commons/httpclient/HttpMethodBase.java

```diff
@@ -1368,6 +1372,84 @@ protected void addCookieRequestHeader(HttpState state, HttpConnection conn)
     }
 
     /**
+     * Put all the cookies in a single header line. 
+     * 
+     * Merge the cookies already present in the request 
+     * with the cookies coming from the state. 
+     * 
+     * @param state the {@link HttpState state} information associated with this method
+     * @param conn the {@link HttpConnection connection} used to execute
+     *        this HTTP method
+     * @param cookies associated with the {@link HttpState state}
+     * 
+     */
+	private void putAllCookiesInASingleHeader(HttpState state, HttpConnection conn,
+			Cookie[] cookies) {
+		
+		LOG.trace("enter putAllCookiesInASingleHeader(HttpState state, HttpConnection conn, Cookie[] cookies)" );
+		
+		String host = getHost(state, conn);
+		CookieSpec matcher = getCookieSpec(state);
+		//use a map to make sure we only have one cookie per name
+		HashMap<String, Cookie> mergedCookies = new HashMap<String, Cookie>();		
+		Header[] cookieheaders = getRequestHeaderGroup().getHeaders("Cookie");
+		for (int i = 0; i < cookieheaders.length; i++) {
+			Header cookieheader = cookieheaders[i];
+			HeaderElement[] headerElements = parseCookieHeaders(cookieheader);
+			for (int j = 0; j < headerElements.length; j++) {
+				Cookie cookie = new Cookie(host,
+						headerElements[j].getName(),
+						headerElements[j].getValue());
+				mergedCookies.put(cookie.getName(),cookie);
+			}
+			// clean the header
+			getRequestHeaderGroup().removeHeader(cookieheader);
+		}
+		//the cookies coming from the state override 
+		//the cookies coming from the request
```

Comment:
I thougth some cookies coming from the initial request could also be present in the state and thus the state should override the request. 
Example (from webgoat)
```
Cookie: has_js=1; JSESSIONID=6C06A25C2B0AECBF33ED5A73A5D2EF05
```

may become 

```
Cookie: has_js=1; JSESSIONID=6C06A25C2B0AECBF33ED5A73A5D2EF05
Cookie: JSESSIONID=5DFA94B903A0063839E0440118808875
```

But you're right that's not the case instead we have this 
```
Cookie: has_js=1
Cookie: JSESSIONID=5DFA94B903A0063839E0440118808875
```
Cookies involved in the session are removed from the request and managed in the state. So we do not override anything.